### ISSUE TITLE
Weird inference for a constructor call to a class with a broken `__new__` method

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#4043]
### Summary

Weird things happen if you create a class with a `__new__` method that's invalidly decorated with `@classmethod` (all `__new__` methods should be staticmethods). ty correctly reports a diagnostic when you try to construct the class -- the `Foo()` call fails at runtime -- but reveals `<class 'Foo'>` as the result of the constructor call. All other type checkers infer `Foo`:

```py
from typing import Self

class Foo:
    @classmethod
    def __new__(cls) -> Self:
        obj = object.__new__(cls)
        obj.attr = 1
        return obj

reveal_type(Foo())
```
